iceoryx_hoofs/include/iceoryx_hoofs/internal/concurrent/periodic_task.hpp

```diff
@@ -123,7 +123,7 @@ class PeriodicTask
     units::Duration m_interval{units::Duration::fromMilliseconds(0U)};
     /// @todo use a refactored posix::Timer object once available
     posix::Semaphore m_stop{posix::Semaphore::create(posix::CreateUnnamedSingleProcessSemaphore, 0U).value()};
-    std::thread m_taskExecutor;
+    cxx::optional<posix::Thread> m_taskExecutor;
```

Comment:
Yes the alias is ok. The problem is that potentially invalid objects (PIO) are viral. If we use an such a PIO in another object, it becomes a PIO itself etc. I am not sure how to deal with this in general and we should strive to avoid them.

But low level APIs may fail and we cannot ignore this.

Ideally we have something during construction which may fail to construct the thread (a `std::thread` would throw but we cannot). But we can terminate or call some error handler.

So we could have a thread handle that cannot be invalid and we assign it only if the construction succeeded. Somehow like this (pseudo code) in the ctor:

```
using maybe_thread_handle_t = cxx::optional<posix::Thread>; // technically no handle yet ...
auto maybeThreadHandle = createThread(...);
if(!maybeThreadHandle) {
   terminate();
}
m_thread = maybeThreadHandle.value();
```

If we do not terminate, the invariant holds that m_thread is a valid handle (unless we reassign and all this but we could avoid this by wrapping sme of the code in a function that either creates a valid handle from a maybe invalid handle or terminates, leaving no other option to create it (only move it later).

Can also be done in the initializer list with e.g. immediately invoked lambdas (not so nice syntactically but this is the price).

Reply:
Exactly. That's why my personal ideal API would be
```cpp
thread_foo_t tft;
thread t { createThread(tft, ...).expect("Aaaaaaaaaaargh!") };
```
or
```cpp
thread_foo_t tft;
thread t { createThread(new thread_foo_t, ...).expect("Aaaaaaaaaaargh!") };
```
And at some point in the future with a better in process memory management just a
```cpp
thread t { createThread(...).expect("Aaaaaaaaaaargh!") };
```

Reply:
@elBoberido @MatthiasKillat the more I think about it the more I am against separating storage from handle in this fashion. The problem is in my opinion ownership and lifetime.

## The solution I prefer, handle and storage combined in one Thread object

First, I assume that the final posix thread wrapper has an interface to set external stack memory so that you can use it like this:
```cxx
cxx::optional<posix::Thread> myThread;
ThreadBuilder().setStackMemory(somePtr, memorySize)
               .create(myThread, someCallback)
               .expect("Aaarrgh!");
```

Or even better that `posix::Thread` comes with an optional template parameter where the stack size is part of the object  so that you can use it like this:
```cxx
cxx::optional<posix::Thread<StackSizeInMb<10>> myThread;
ThreadBuilder().create(myThread, someCallback)
               .expect("Aaarrgh!");
```

When one would like to store the posix thread somewhere else this is easy peasy, just use
```
using threadHandle_t = cxx::optional<posix::Thread<StackSizeInMb<10>>
threadHandle_t * myThread = new threadHandle_t();
// or
threadHandle_t * memory = new (allocateSomeMemory(sizeof(threadHandle_t))) threadHandle_t();
```

The big advantage is, the thread owns the stack memory and ensures its lifetime.

## Solution with separate storage and handle

But when we would pursue an approach which separates storage from handle similar to this:
```cxx
threadStorage_t threadStorage;
threadHandle_t threadHandle = ThreadBuilder().create(...).expect("Aarrgh!");
```
We have the following problems where we do not have an answer to and which increases the code complexity when solving them:
1. Lifetime, what happens when `threadStorage` goes out of scope before `threadHandle`
   * undefined behavior, this can be ugly
   * `threadStorage` knows that it is in use
     * calls thread_join - which may surprises the users and can lead easily to deadlocks and would also require that the `threadHandle` knows the `threadStorage` so that it can release the memory when they go out of scope in the right order.
     * calls terminate - we had this in the first waitset and this just makes the code unnecessarily unstable in my opinion
2. Ownership, those two objects are tightly coupled together and I would strongly refrain from using `threadStorage` for different threads but it would be legal. I would be fine with restarting the same thread over and over again if this would be a use case but completely different threads would be a pain. In the end I would always couple those together like
```cxx
struct ThreadStuff {
  threadStorage_t storage;
  threadHandle_t handle;
};
```
3. Ownership the second, what happens when a second thread is created in the `threadStorage`? Terminate, thread_join or thread_exit of the existing thread, undefined behavior? Since `threadHandle` is independent of `threadStorage` we cannot check if `threadHandle` is set since a user maybe initialized a different handle. A solution would be then to check if the `threadStorage` is already in use. So it would not be as easily usable as suggested, one would have to write stuff like:
```cxx
threadStorage_t threadStorage;
threadHandle_t threadHandle;
//...
if ( threadStorage.isNotInUse() ) {
  threadHandle = ThreadBuilder().create(...).expect("Aaargh!");
} else {
  errorHandler(THREAD_STORAGE_NOT_AVAILABLE);
}
```

## Summary

In my opinion separating thread storage from handle has lifetime and ownership issues which can be solved but make the thread implementation much more complex and the usage more error prone and also unnecessary complex.

I also have a hard time seeing the advantage of separating storage from handle except that the handle would be movable which could be also solved by introducing a semaphore into the thread which signals when the `startRoutine` has copied the callable.
But we have no where in iceoryx the use case that the thread must be movable. And a user could solve this easily by storing it in the heap, and we could solve this easily by adding an allocator to our posh runtime (or somewhere else) where we already have plans for.

Reply:
@MatthiasKillat @elBoberido @elfenpiff Would you agree to merge the current changes (once the CI is green) as a first replacement for the std::thread occurrences? After making it work on Windows and Mac and replacing std::thread, I would continue with implementing a way to set the stack address and reconsider all your comments and suggestions. I think the current implementation doesn't hamper further design approaches and should be easily expandable/adaptable.

Reply:
@FerdinandSpitzschnueffler agreed

@elfenpiff these are valid concerns but the two approaches are not that different.

Considering that something like `cxx::optional<posix::Thread<StackSizeInMb<10>> myThread;` won't work since most probably there is not enough stack on your current thread to put the storage of the new thread onto that stack, all of your concerns regarding the separate storage from the handle can be solved the same way you propose to solve the stack problem when it is part of the handle.

```cpp
iox::cxx::unique_ptr<threadStorage_t> ts {new threadStorage_t};
auto myThread = ThreadBuilder(std::move(ts))
           .create(someCallback)
           .expect("Aaarrgh!");
```

The nice thing about this is that we do not need to check if the `optional` contains a value, since there is no `optional`. Of course we can argue that we abort when `create` fails but what about this

```cpp
cxx::optional<posix::Thread> myThread;
ThreadBuilder().setStackMemory(somePtr, memorySize)
               .create(myThread, someCallback)
               .or_else("Aaarrgh!");
...
myThread.reset();
...
myThread.getName(); // kaboom
```

Reply:
@elfenpiff @elBoberido @MatthiasKillat I summarized and linked the main discussions in #1365 so that nothing is forgotten.

Reply:
@FerdinandSpitzschnueffler each type of API has it drawbacks and it all boils down to which drawbacks make us less uncomfortable

Reply:
@elBoberido Agreed. That's why I added the arguments for both to the issue.